src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -2232,15 +2232,47 @@ fn ambiguity_error(
     emit_error: bool,
 ) -> bool {
     let mut descrs = FxHashSet::default();
-    let kinds = candidates
+    // proc macros can exist in multiple namespaces at once,
+    // so we need to compare DefIds to remove
+    // the candidite in the fn namespace
+    let mut possible_proc_macro_id = None;
+    let is_proc_macro_crate = cx.tcx.crate_types() == &[CrateType::ProcMacro];
+    let mut kinds = candidates
         .iter()
-        .map(
-            |(res, def_id)| {
-                if let Some(def_id) = def_id { Res::from_def_id(cx.tcx, *def_id) } else { *res }
-            },
-        )
-        .filter(|res| descrs.insert(res.descr()))
+        .map(|(res, def_id)| {
+            let r =
+                if let Some(def_id) = def_id { Res::from_def_id(cx.tcx, *def_id) } else { *res };
+            if is_proc_macro_crate && let Res::Def(DefKind::Macro(_), id) = r {
+                possible_proc_macro_id = Some(id);
+            }
+            r
+        })
         .collect::<Vec<_>>();
+    // in order to properly dedup proc macros,
+    // we have to do it in two passes,
+    // completing the full traversal to find
+    // the possible duplicate in the macro namespace,
+    // then another full traversal to eliminate the
+    // candidite in the fn namespace.
+    // thus, we have to do an iteration after
+    // collection is finished.
+    //
+    // as an optimization, we only deduplicate if we're in a proc-macro crate,
+    // and only if we already found something that looks like a proc macro.
+    if is_proc_macro_crate && let Some(macro_id) = possible_proc_macro_id {
+        kinds.retain(|res| {
+            if let Res::Def(DefKind::Fn, fn_id) = res
+                && macro_id == *fn_id
+            {
+                false
+            } else {
+                true
+            }
```

Comment:
I think this could work:

```suggestion
            !matches!(res, Res::Def(DefKind::Fn, fn_id) if macro_id == *fn_id)
```